### PR TITLE
[JENKINS-49853] Provide a cloud-friendly installation for AWS

### PR DESCRIPTION
**Pull request by batmat**
[JENKINS-49853](https://issues.jenkins-ci.org/browse/JENKINS-49853)

This change provides an auto-configured version of Jenkins Essentials for AWS. 
* This will auto-configure the [EC2 plugin](https://plugins.jenkins.io/ec2) to be able to dynamically provision EC2 agents VMs
* This will also auto-configure the [Artifact Manager on S3](https://plugins.jenkins.io/artifact-manager-s3) plugin so that artifacts are not stored anymore on the master's FileSystem, but directly in S3.

To create the whole setup, we use CloudFormation.

I have added a README in the `environments` directory for the AWS flavor.

cc @jglick & @oleg-nenashev who I coudn't add directly as reviewers.

Any feedback is welcome.

Demo video:
https://www.youtube.com/watch?v=966OiFW2gow

### CODE REVIEW

File: distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json [hunk 1]
```diff
@@ -0,0 +1,165 @@
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Parameters" : {
+    "InstanceTypeParameter" : {
+      "Type" : "String",
+      "Default" : "t2.large",
+      "AllowedValues" : ["t2.micro", "m1.small", "m1.large", "t2.large"],
```
Comment (batmat) [inline distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: I should change that. Allowing a `t2.micro` probably makes no sense. But then, should I actually list *all* allowed values, not sure.

File: distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        ],
+        "UserData": {
+          "Fn::Base64" : {
+            "Fn::Join" : ["", [
+              "#!/bin/bash -v\n",
+              "touch blah\n",
+              "echo ",{"Ref":"S3BucketForArtifactManager"}, "> BUCKET_NAME\n",
+              "echo \"", { "Ref": "PrivateKey"}, "\" > ssh-agents-private-key\n",
+              "#chmod og-rw ssh-agents-private-key\n",
+              "#docker run -e EVERGREEN_ENDPOINT=http://evergreen.jenkins.io.batmat.net:8080 -d -p 8080:8080 -p 50000:50000 batmat/evergreen:docker-cloud\n",
```
Comment (batmat) [inline distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: note to myself: clean this up...

File: distribution/environments/build-or-push-flavors.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+set -euo pipefail
+
+IMAGE_NAME=jenkins/evergreen
+
+# docker build -t $IMAGE_NAME:$currentDir .
+# docker push $IMAGE_NAME:$currentDir
```
Comment (batmat) [inline distribution/environments/build-or-push-flavors.sh]: remove this

File: distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json (hunk 1 above)
Comment (batmat) [reply distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: I tend to agree, though we also know that running a master on 512MB of RAM would very likely be a recipe for failure. I couldn't find the predefined type for instance types, though. Care to point me to it? I agree it looks weird it's not defined, but I cannot find it in https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-specific-parameter-types @mandie722

**Comment by batmat** [comment]
Failed apparently because of timeout.

**Comment by batmat** [comment]
The whole `make check` takes 14 minutes on my machine. Seems way longer in CI, 37 minutes already and counting right now :open_mouth: 
image

File: distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json (hunk 1 above)
Comment (mandie722) [reply distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: It's under the Elastic section: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html

Comment (batmat) [reply distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: @mandie722 seems more something usable when calling remote API, here XML AIUI when going on the higher levels of that subdocumentation? What type/value would I use here in e cloudformation template?

**Review by mandie722 — APPROVED** [review]

File: distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                  "Resource": "*"
+              }]
+            }
+          }, {
+            "PolicyName": "S3ArtifactManagerPolicy",
+            "PolicyDocument": {
+              "Version": "2012-10-17",
+              "Statement": [
+                  {
+                      "Sid": "TodoRefineSecurityALot",
```
Comment (batmat) [inline distribution/environments/aws-ec2-cloud/CloudFormation/cloudformation-template.json]: Filed https://issues.jenkins-ci.org/browse/JENKINS-52342 to follow up on this.

File: distribution/environments/aws-ec2-cloud/config/as-code/artifact-manager-s3.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+---
+# TODO: file JIRAs to make this classifed (and for other used plugins too)
+unclassified:
+  s3blobstoreconfig:
+    # TODO: create through CloudFormation
+    container: "${ARTIFACT_MANAGER_S3_BUCKET_NAME}"
+    # FIXME: not a sensible default for everyone
+    region: "us-east-1"
```
Comment (jglick) [inline distribution/environments/aws-ec2-cloud/config/as-code/artifact-manager-s3.yaml]: Note that this is invalid as of https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/67.

File: distribution/environments/aws-ec2-cloud/config/as-code/artifact-manager-s3.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+---
+# TODO: file JIRAs to make this classifed (and for other used plugins too)
+unclassified:
```
Comment (jglick) [inline distribution/environments/aws-ec2-cloud/config/as-code/artifact-manager-s3.yaml]: This too.

**Pull request merged**